### ISSUE TITLE
Number of feature after block 3

### ISSUE THREAD

**Issue by John1231983** [NanqingD/DeepLabV3-Tensorflow#6]
Hello, I checked your implementation in ASPP model with resnet 101. I foubd that your number of feature map after block 3 ( starts use dilation with rate 2) and in the ASPP may be not correct. If I understand, it must be 1024 for block 3, 2048 for block 4 and 2048×5 for ASPP, instead of 256 in your code. Am I right?

**Comment by NanqingD** [comment]
Hi, John. Thanks for the reply.
![resnet](https://user-images.githubusercontent.com/16928463/34115118-a5c85514-e3e2-11e7-9224-c04b19e5611c.png)
I try to reimplement this.
```
with tf.variable_scope('block4', [net]) as sc:
   base_depth = 512
   for i in range(3):
      with tf.variable_scope('unit_%d' % (i + 1), values=[net]):
         net = bottleneck_hdc(net, depth=base_depth * 4,depth_bottleneck=base_depth, stride=1, rate=2)
         net = slim.utils.collect_named_outputs(end_points_collection, sc.name, net)
```
```
def bottleneck_hdc(inputs,
               depth,
               depth_bottleneck,
               stride,
               rate=1,
               outputs_collections=None,
               scope=None,
               use_bounded_activations=False):
  """Hybrid Dilated Convolution Bottleneck.
  Multi_Grid = (1,2,4)
  See Understanding Convolution for Semantic Segmentation.
  When putting together two consecutive ResNet blocks that use this unit, one
  should use stride = 2 in the last unit of the first block.
  Args:
    inputs: A tensor of size [batch, height, width, channels].
    depth: The depth of the ResNet unit output.
    depth_bottleneck: The depth of the bottleneck layers.
    stride: The ResNet unit's stride. Determines the amount of downsampling of
      the units output compared to its input.
    rate: An integer, rate for atrous convolution.
    outputs_collections: Collection to add the ResNet unit output.
    scope: Optional variable_scope.
    use_bounded_activations: Whether or not to use bounded activations. Bounded
      activations better lend themselves to quantized inference.
  Returns:
    The ResNet unit's output.
  """
  with tf.variable_scope(scope, 'bottleneck_v1', [inputs]) as sc:
    depth_in = slim.utils.last_dimension(inputs.get_shape(), min_rank=4)
    if depth == depth_in:
      shortcut = resnet_utils.subsample(inputs, stride, 'shortcut')
    else:
      shortcut = slim.conv2d(
          inputs,
          depth, [1, 1],
          stride=stride,
          activation_fn=tf.nn.relu6 if use_bounded_activations else None,
          scope='shortcut')

    residual = slim.conv2d(inputs, depth_bottleneck, [1, 1], stride=1, 
        rate=rate*1, scope='conv1')
    residual = resnet_utils.conv2d_same(residual, depth_bottleneck, 3, stride,
                                        rate=rate*2, scope='conv2')
    residual = slim.conv2d(residual, depth, [1, 1], stride=1, rate=rate*4,
                           activation_fn=None, scope='conv3')

    if use_bounded_activations:
      # Use clip_by_value to simulate bandpass activation.
      residual = tf.clip_by_value(residual, -6.0, 6.0)
      output = tf.nn.relu6(shortcut + residual)
    else:
      output = tf.nn.relu(shortcut + residual)

    return slim.utils.collect_named_outputs(outputs_collections,
                                            sc.name,
                                            output)
```

**Comment by NanqingD** [comment]
But nice spot, I do have a bug for block 4, it should be multigrid, so `bottleneck` -> `bottleneck_hdc`.

**Comment by John1231983** [comment]
Thanks for your comment. I think you are implementing the cascade model. For ASPP model, how about the feature map `256` in this [line](https://github.com/NanqingD/DeepLabV3-Tensorflow/blob/master/libs/nets/deeplabv3.py#L146)? I think it must be the same number of feature map of the previous block (block 7 in your case).

**Comment by NanqingD** [comment]
Hi, John, the following is from the paper.

> In the end, our improved ASPP consists of (a) one 1×1 convolution
and three 3 × 3 convolutions with rates = (6, 12, 18) when
output stride = 16 (all with **256** filters and batch normalization),
and (b) the image-level features, as shown in Fig. 5.
Note that the rates are doubled when output stride = 8. 

My understanding is that 1) the author compress number of filters from 2048 to 256 to control the complexity of the network, 2) the ASPP module one 1x1 and three 3x3 conv, and the image pooling are in parallel. My evidence is that the author said "our improved ASPP consists of", since the old ASPP in DeepLabV2 is parallel style, so I assume here a) and b) are in parallel. 

Correct me if I am wrong since I am exploring this model too.

**Comment by John1231983** [comment]
Sorry, I did not find the sentence before. You are totally correct. You are only missing batchnorm in ASPP, but you mentioned it in the README. Thanks for help me clear understand it. For the second point, I also agree with you that global pooling must be parallel with other convolution. I asked the point because I found someone implemented the deep lab and located the pool after concatenation

Finally, we can draw the deeplab-v3 as (just simple assume the input size as `224x224` instead of `513x513`)

```
Input (224x224x3)->conv (7,2)+pool(2)->Output (56x56x64): out_stride=4
Input (56x56x64)->block1->Output (56x56x256): out_stride=4
Input (56x56x256)->block2->Output (28x28x512): out_stride=8
Input (28x28x512)->block3->Output (14x14x1024): out_stride=16
Input (14x14x1024)->block4+rate=[1,2,4]->Output (14x14x2048): out_stride=16
Input (14x14x1024)->block5+rate=[1,2,4]->Output (14x14x2048): out_stride=16
Input (14x14x1024)->block6+rate=[1,2,4]->Output (14x14x2048): out_stride=16
Input (14x14x1024)->block7+rate=[1,2,4]->Output (14x14x2048): out_stride=16
#ASPP model
Input (14x14x2048)->conv (1x1)+ BN->Output (14x14x256) (1)
Input (14x14x2048)->conv (3x3)+rate=6+ BN->Output (14x14x256) (2)
Input (14x14x2048)->conv (3x3)+rate=12+ BN->Output (14x14x256) (3)
Input (14x14x2048)->conv (3x3)+rate=18+ BN->Output (14x14x256) (4)
Input (14x14x2048)->average pooling(k=2,s=2)->Output (14x14x2048) (5)
#Classification
Concat(1,2,3,4,5)->conv(1,1)+ BN->Output (14x14x256) 
Input(14x14x256)->conv(1,1)->Output (14x14x21) #number of classes 
```
 
Note that, the Fig.5 in the paper only have 4 blocks, while the final result he used 7 blocks (extra 3 blocks). Please look at the network and correct if I was wrong. Currently, I achieve `69.9%` with crop size 513, 4 blocks, without batchnorm, (1,2,4) grid size on PASCAL VOC

**Comment by NanqingD** [comment]
Hi, John, the batch_norm is added automatically by 
`with slim.arg_scope(resnet_arg_scope(weight_decay=args.weight_decay)):`.
